### PR TITLE
Admiral 47 exactly reproducible experiments

### LINKED ISSUE #47: Exactly reproducible random environments
Every experiment must be reproducible or the environments have not much worth for research because we can't compare runs properly. There needs to be a random number generator which gets a seed for each environment and agent so that one can run identical experiments if they choose to. The particle environment can be used as an example--it sets the random number generator for the action spaces as well. This is a huge thing in RL.

### PR DESCRIPTION
**Pull request by rusu24edward**
Resolves #47

### CODE REVIEW

**Comment by rusu24edward** [comment]
@glatt1 Each agent can be given its own seed. Let me know if this is what you had in mind about reproducibility.

File: admiral/envs/components/agent.py [hunk 1]
```diff
@@ -39,8 +40,14 @@ def __init__(self, action_space=None, **kwargs):
         self.action_space = {} if action_space is None else action_space
 
     def finalize(self, **kwargs):
+        """
+        Wrap all the action spaces with a Dict and seed it if the agent was
+        created with a seed.
+        """
         super().finalize(**kwargs)
         self.action_space = Dict(self.action_space)
+        if self.seed is not None:
```
Comment (glatt1) [inline admiral/envs/components/agent.py:49]: I don't think the 'if ...' is necessary. It makes sense to seed every time and the underlying function can handle 'None' values.

File: admiral/envs/components/agent.py [hunk 2]
```diff
@@ -56,8 +63,14 @@ def __init__(self, observation_space=None, **kwargs):
         self.observation_space = {} if observation_space is None else observation_space
 
     def finalize(self, **kwargs):
+        """
+        Wrap all the observation spaces with a Dict and seed it if the agent was
+        created with a seed.
+        """
         super().finalize(**kwargs)
         self.observation_space = Dict(self.observation_space)
+        if self.seed is not None:
```
Comment (glatt1) [inline admiral/envs/components/agent.py:72]: Same as above: I don't think the 'if ...' is necessary. It makes sense to seed every time and the underlying function can handle 'None' values.

**Comment by glatt1** [comment]
> @glatt1 Each agent can be given its own seed. Let me know if this is what you had in mind about reproducibility.

yes, this is what I had in mind. Very nice!

Is the ID always an integer? In the case no seed is set, would it make sense to just use the ID as seed? That way it's always seeded and we can be sure no two agents have the same seed (if we don't want that specifically).

**Comment by rusu24edward** [comment]
> Is the ID always an integer? In the case no seed is set, would it make sense to just use the ID as seed? That way it's always seeded and we can be sure no two agents have the same seed (if we don't want that specifically).

What is the benefit and possible side effects of having the agent always seeded?

**Comment by glatt1** [comment]
> > Is the ID always an integer? In the case no seed is set, would it make sense to just use the ID as seed? That way it's always seeded and we can be sure no two agents have the same seed (if we don't want that specifically).
> 
> What is the benefit and possible side effects of having the agent always seeded?

@rusu24edward It's already always seeded now with the last changes but If the seed is None it will be a random seed. We can control that by adding basically any number and then make experiments reproducible without giving a specific seed (not the case with None). Sometimes we want the seed to be the same for different agents so they have the same starting conditions but most of the time different seeds per agent is better. So, if we fix the seed for general purpose to the ID we just need to set a seed if we want agents with the same starting conditions in a single experiment.
Not sure if there are any downsides. RLlib addressed this in this PR: https://github.com/ray-project/ray/pull/5197

**Comment by rusu24edward** [comment]
> So, if we fix the seed for general purpose to the ID we just need to set a seed if we want agents with the same starting conditions in a single experiment.

I hesitate to fix the seed for an agent by default. I think it would be better if the default behavior is not seeded (i.e. random seed determined by the system). If I'm reading the rllib code correctly, I believe this is the pattern they follow as well. For us, seed is set if and only if the user explicitly makes the agent with a seed. This way we still have reproducibility if the user purposefully requests it.

**Review by glatt1 — APPROVED** [review]
Ok, let's leave it as is.

**Pull request merged**